src/distributions/slice.rs

```diff
@@ -115,3 +117,22 @@ impl core::fmt::Display for EmptySlice {
 
 #[cfg(feature = "std")]
 impl std::error::Error for EmptySlice {}
+
+/// Note: the `String` is potentially left with excess capacity; optionally the
+/// user may call `string.shrink_to_fit()` afterwards.
+#[cfg(feature = "alloc")]
+impl<'a> super::DistString for Slice<'a, char> {
+    fn append_string<R: crate::Rng + ?Sized>(&self, rng: &mut R, string: &mut String, len: usize) {
+        let max_char_len = self
+            .slice
+            .iter()
+            .try_fold(1, |max_len, char| {
+                // When the current max_len is 4, the result max_char_len will be 4.
+                Some(max_len.max(char.len_utf8())).filter(|len| *len < 4)
+            })
+            .unwrap_or(4);
+
+        string.reserve(max_char_len * len);
+        string.extend(self.sample_iter(rng).take(len))
```

Comment:
If the slice is large, this could take significant time (and the only purpose is to potentially reduce the max-char-len used for space reservation below 4). I suggest only iterating when the length is under some bound (1000 maybe?).

Further, a slice could contain one 4-byte character but mostly 1-byte chars so this could massively over-reserve (perhaps significant if `len` is large). It isn't obvious what the best approach is (or how to test — it could be very context dependent — perhaps try an approach which isn't terrible anywhere rather than to perfectly optimise).

I mean it's arguable whether just reserving `len` bytes would be better. Or you could sample `len/4`, check the length and repeat... but probably over-complex.

Reply:
What is the goal here? How many allocations are acceptable? We could always have conservative lower limit and let `extend` deal with the additional allocations. If we reserve the minimal `len/4`, this should result in at most 3 allocations, right? I think this is good enough for a "general-purpose" implementation.

Are there use cases where the slice is large? Typically, I would expect the "alphabet" to be small, unless it is abused for weighted sampling, for which we have better implementations. Another use case could be "valid Unicode except a few characters", but for this, rejection sampling should be used.

Reply:
> Typically, I would expect the "alphabet" to be small

As would I, but it *could* be large and this is easy to test. The simplest solution would be to measure max-char-size only where the alphabet is reasonably small, e.g. under 200 items.